lucille-core/src/main/java/com/kmwllc/lucille/stage/ApplyJavascript.java

```diff
@@ -0,0 +1,432 @@
+      throw new StageException("Can only specify either script path or script.");
+    }
+
+    if (inlineScript != null) {
+      try {
+        this.source = Source.newBuilder("js", inlineScript, "<inline>").build();
+      } catch (Exception e) {
+        throw new StageException("Failed to build inline JavaScript source.", e);
+      }
+    } else {
+      try (Reader reader = FileContentFetcher.getOneTimeReader(scriptPath, config)) {
+        this.source = Source.newBuilder("js", reader, scriptPath).build();
+      } catch (Exception e) {
+        throw new StageException("Failed to read JavaScript from '" + scriptPath + "'.", e);
+      }
+    }
+
+    try {
+      this.engine = Engine.newBuilder()
+          .option("engine.WarnInterpreterOnly", "false")
+          .option("log.level", "OFF")
+          .build();
+
+      this.context = Context.newBuilder("js")
+          .engine(engine)
+          .allowAllAccess(true)
+          .build();
+    } catch (Exception e) {
+      throw new StageException("Failed to initialize JavaScript context.", e);
+    }
+  }
+
+  @Override
+  public void stop() throws StageException {
+    if (context != null) {
+      context.close();
+    }
+
+    if (engine != null) {
+      engine.close();
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    try {
+      context.getBindings("js").putMember("doc", JsDocProxy.root(doc));
+      context.getBindings("js").putMember("rawDoc", doc);
+      context.eval(source);
+      return null;
+    } catch (Exception e) {
+      throw new StageException("JavaScript failed for doc '" + doc.getId() + "' using '" + scriptPath + "'.", e);
+    }
+  }
+
+  // Proxy implementation exposed to JS
+
+  private static final class JsDocProxy implements ProxyObject {
+    private static final JsonNodeFactory JSON = JsonNodeFactory.instance;
+
+    private final Document doc;
+    private final String path; // null for root, otherwise "a.b"
+
+    private JsDocProxy(Document doc, String path) {
+      this.doc = doc;
+      this.path = path;
+    }
+
+    static JsDocProxy root(Document d) {
+      return new JsDocProxy(d, null);
+    }
+
+    private boolean isRoot() {
+      return path == null;
+    }
+    private String join(String key) {
+      return isRoot() ? key : (path + "." + key);
+    }
+
+    @Override
+    public Object getMember(String key) {
+      return isRoot() ? getRootMember(key) : getNestedMember(key);
+    }
+
+    private Object getRootMember(String key) {
+      if (!doc.has(key)) {
+        return null;
+      }
+
+      JsonNode json = doc.getJson(key);
+
+      if (json != null) {
+        if (json.isNull()) return null;
+        if (json.isArray()) return jsonArrayToProxyArray(json, key);
+        if (json.isContainerNode()) return new JsDocProxy(doc, key);
+        return jsonNodeToJsValue(json);
+      }
+
+      Object raw = doc.asMap().get(key);
```

Comment:
how could we ever reach this line?
if doc.has(key) then doc.getJson(key) will always return non-null for JsonDocument -- looking at the implementation it just returns data.get(name)

so maybe this line can only be reached with HashMapDocument?
I'm going to suggest we limit ApplyJavascript to being used with JsonDocument because HashMapDocument is not really in use in the codebase and should be deprecated

would that save us some code here?